File: backend/src/main/java/com/rgt/restaurant/service/CategoryService.java
package com.rgt.restaurant.service;

import com.rgt.restaurant.model.Category;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

@Service
@Slf4j
@RequiredArgsConstructor
public class CategoryService {
    private final Map<String, Category> categories = new ConcurrentHashMap<>();
    private final MenuCategoryService menuCategoryService;

    public Category createCategory(String name, String description) {
        Category category = new Category(UUID.randomUUID().toString(), name, description);
        categories.put(category.getId(), category);
        return category;
    }

    public List<Category> getAllCategories() {
        return new ArrayList<>(categories.values());
    }

    public Category getCategory(String id) {
        return categories.get(id);
    }

    public void deleteCategory(String id) {
        if (!categories.containsKey(id)) {
            throw new IllegalArgumentException("Category not found");
        }
        
        if (menuCategoryService.categoryHasMenus(id)) {
            throw new IllegalStateException("Cannot delete category with existing menu items");
        }
        
        categories.remove(id);
    }

    public Category updateCategory(String id, String name, String description) {
        Category category = categories.get(id);
        if (category == null) {
            log.warn("Attempted to update non-existent category {}", id);
            throw new IllegalArgumentException("Category not found");
        }
        
        category.setName(name);
        category.setDescription(description);
        categories.put(id, category);
        log.info("Category updated - ID: {}, Name: {}", id, name);
        return category;
    }
} 

File: frontend/src/services/categoryService.ts
import axios from 'axios';
import type { Category } from '../types/Menu';

const API_URL = 'http://localhost:8080/api';

export const categoryService = {
  createCategory: async (name: string, description?: string): Promise<Category> => {
    try {
      const response = await axios.post(`${API_URL}/category`, {
        name,
        description
      });
      return response.data;
    } catch (error) {
      throw new Error('Failed to create category');
    }
  },

  getAllCategories: async (): Promise<Category[]> => {
    try {
      const response = await axios.get(`${API_URL}/category`);
      return response.data;
    } catch (error) {
      throw new Error('Failed to fetch categories');
    }
  },

  deleteCategory: async (id: string): Promise<void> => {
    try {
      await axios.delete(`${API_URL}/category/${id}`);
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.status === 409) {
        throw new Error('Cannot delete category with existing menu items');
      }
      throw new Error('Failed to delete category');
    }
  },

  updateCategory: async (id: string, name: string, description?: string): Promise<Category> => {
    try {
      const response = await axios.put(`${API_URL}/category/${id}`, {
        name,
        description
      });
      return response.data;
    } catch (error) {
      throw new Error('Failed to update category');
    }
  }
}; 